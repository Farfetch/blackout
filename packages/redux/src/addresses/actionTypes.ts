/**
 * Action type dispatched when the fetch prediction request fails.
 */
export const FETCH_ADDRESS_PREDICTIONS_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_FAILURE';
/**
 * Action type dispatched when the fetch prediction request starts.
 */
export const FETCH_ADDRESS_PREDICTIONS_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_REQUEST';
/**
 * Action type dispatched when the fetch prediction request succeeds.
 */
export const FETCH_ADDRESS_PREDICTIONS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_SUCCESS';

/**
 * Action type dispatched when the fetch prediction details request fails.
 */
export const FETCH_ADDRESS_PREDICTION_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_FAILURE';
/**
 * Action type dispatched when the fetch prediction details request starts.
 */
export const FETCH_ADDRESS_PREDICTION_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_REQUEST';
/**
 * Action type dispatched when the fetch prediction details request succeeds.
 */
export const FETCH_ADDRESS_PREDICTION_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_SUCCESS';

/**
 * Action type dispatched when resetting the address prediction.
 */
export const RESET_ADDRESS_PREDICTIONS =
  '@farfetch/blackout-redux/RESET_ADDRESS_PREDICTIONS';
