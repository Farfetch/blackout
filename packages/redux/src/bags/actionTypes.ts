/**
 * @module bags/actionTypes
 * @category Bags
 * @subcategory Action Types
 */

/** Action type dispatched when the add bag item request fails. */
export const ADD_BAG_ITEM_FAILURE =
  '@farfetch/blackout-redux/ADD_BAG_ITEM_FAILURE';
/** Action type dispatched when the add bag item request starts. */
export const ADD_BAG_ITEM_REQUEST =
  '@farfetch/blackout-redux/ADD_BAG_ITEM_REQUEST';
/** Action type dispatched when the add bag item request succeeds. */
export const ADD_BAG_ITEM_SUCCESS =
  '@farfetch/blackout-redux/ADD_BAG_ITEM_SUCCESS';

/** Action type dispatched when the remove item from bag request fails. */
export const REMOVE_BAG_ITEM_FAILURE =
  '@farfetch/blackout-redux/REMOVE_BAG_ITEM_FAILURE';
/** Action type dispatched when the remove item from bag request starts. */
export const REMOVE_BAG_ITEM_REQUEST =
  '@farfetch/blackout-redux/REMOVE_BAG_ITEM_REQUEST';
/** Action type dispatched when the remove item from bag request succeeds. */
export const REMOVE_BAG_ITEM_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_BAG_ITEM_SUCCESS';

/** Action type dispatched when the fetch bag request fails. */
export const FETCH_BAG_FAILURE = '@farfetch/blackout-redux/FETCH_BAG_FAILURE';
/** Action type dispatched when the fetch bag request starts. */
export const FETCH_BAG_REQUEST = '@farfetch/blackout-redux/FETCH_BAG_REQUEST';
/** Action type dispatched when the fetch bag request succeeds. */
export const FETCH_BAG_SUCCESS = '@farfetch/blackout-redux/FETCH_BAG_SUCCESS';

/** Action type dispatched when the reset bag state occurs. */
export const RESET_BAG_STATE = '@farfetch/blackout-redux/RESET_BAG_STATE';
/** Action type dispatched when the reset bag state entities occurs. */
export const RESET_BAG_ENTITIES = '@farfetch/blackout-redux/RESET_BAG_ENTITIES';

/** Action type dispatched when the update bag request fails. */
export const UPDATE_BAG_ITEM_FAILURE =
  '@farfetch/blackout-redux/UPDATE_BAG_ITEM_FAILURE';
/** Action type dispatched when the update bag request starts. */
export const UPDATE_BAG_ITEM_REQUEST =
  '@farfetch/blackout-redux/UPDATE_BAG_ITEM_REQUEST';
/** Action type dispatched when the update bag request succeeds. */
export const UPDATE_BAG_ITEM_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_BAG_ITEM_SUCCESS';
