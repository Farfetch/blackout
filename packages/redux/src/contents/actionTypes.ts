/**
 * Action type dispatched when the fetch commerce pages request fails.
 */
export const FETCH_COMMERCE_PAGES_FAILURE =
  '@farfetch/blackout-redux/FETCH_COMMERCE_PAGES_FAILURE';
/**
 * Action type dispatched when the fetch commerce pages request starts.
 */
export const FETCH_COMMERCE_PAGES_REQUEST =
  '@farfetch/blackout-redux/FETCH_COMMERCE_PAGES_REQUEST';
/**
 * Action type dispatched when the fetch commerce pages request succeeds.
 */
export const FETCH_COMMERCE_PAGES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_COMMERCE_PAGES_SUCCESS';

/**
 * Action type dispatched when the fetch contents request fails.
 */
export const FETCH_CONTENTS_FAILURE =
  '@farfetch/blackout-redux/FETCH_CONTENTS_FAILURE';
/**
 * Action type dispatched when the fetch contents request starts.
 */
export const FETCH_CONTENTS_REQUEST =
  '@farfetch/blackout-redux/FETCH_CONTENTS_REQUEST';
/**
 * Action type dispatched when the fetch contents request succeeds.
 */
export const FETCH_CONTENTS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_CONTENTS_SUCCESS';

/**
 * Action type dispatched when the fetch content types request fails.
 */
export const FETCH_CONTENT_TYPES_FAILURE =
  '@farfetch/blackout-redux/FETCH_CONTENT_TYPES_FAILURE';
/**
 * Action type dispatched when the fetch content types request starts.
 */
export const FETCH_CONTENT_TYPES_REQUEST =
  '@farfetch/blackout-redux/FETCH_CONTENT_TYPES_REQUEST';
/**
 * Action type dispatched when the fetch content types request succeeds.
 */
export const FETCH_CONTENT_TYPES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_CONTENT_TYPES_SUCCESS';

/** Action type dispatched when the fetch content pages request fails. */
export const FETCH_CONTENT_PAGES_FAILURE =
  '@farfetch/blackout-redux/FETCH_CONTENT_PAGES_FAILURE';

/** Action type dispatched when the fetch content pages request starts. */
export const FETCH_CONTENT_PAGES_REQUEST =
  '@farfetch/blackout-redux/FETCH_CONTENT_PAGES_REQUEST';

/** Action type dispatched when the fetch content pages request succeeds. */
export const FETCH_CONTENT_PAGES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_CONTENT_PAGES_SUCCESS';

/**
 * Action type dispatched when the fetch SEO request fails.
 */
export const FETCH_SEO_METADATA_FAILURE =
  '@farfetch/blackout-redux/FETCH_SEO_METADATA_FAILURE';
/**
 * Action type dispatched when the fetch SEO request starts.
 */
export const FETCH_SEO_METADATA_REQUEST =
  '@farfetch/blackout-redux/FETCH_SEO_METADATA_REQUEST';
/**
 * Action type dispatched when the fetch SEO request succeeds.
 */
export const FETCH_SEO_METADATA_SUCCESS =
  '@farfetch/blackout-redux/FETCH_SEO_METADATA_SUCCESS';

/**
 * Action type dispatched when reset contents.
 */
export const RESET_CONTENTS = '@farfetch/blackout-redux/RESET_CONTENTS';
