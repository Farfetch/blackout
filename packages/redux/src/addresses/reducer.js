/**
 * @module addresses/reducer
 * @category Addresses
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createReducerWithResult, reducerFactory } from '../helpers';
import { LOGOUT_SUCCESS } from '@farfetch/blackout-redux/authentication/actionTypes';
import get from 'lodash/get';
import omit from 'lodash/omit';
import produce from 'immer';

const INITIAL_STATE = {
  error: null,
  isLoading: false,
  result: null,
  predictions: {
    result: null,
    error: null,
    isLoading: false,
  },
  predictionDetails: {
    result: null,
    error: null,
    isLoading: false,
  },
  addresses: {
    error: null,
    isLoading: false,
  },
  address: {
    error: {},
    isLoading: {},
  },
  addressSchema: {
    error: null,
    isLoading: false,
  },
  /* Used for operations related with the default address that
    have a result associated, such as getting the default contact address */
  defaultAddressDetails: {
    error: null,
    isLoading: false,
    result: null,
  },
};

export const getDefaultAddress = (addressesList, prop) => {
  for (const key in addressesList) {
    if (addressesList[key] && addressesList[key][prop]) {
      return addressesList[key];
    }
  }
  return null;
};

const result = (state = INITIAL_STATE.result, action = {}) => {
  switch (action.type) {
    case actionTypes.FETCH_ADDRESSES_SUCCESS:
      return action.payload.result;
    case actionTypes.CREATE_ADDRESS_SUCCESS:
      return [...state, action.meta.addressId];
    case actionTypes.REMOVE_ADDRESS_SUCCESS: {
      return (
        state && state.filter(addressId => addressId !== action.meta.addressId)
      );
    }
    default:
      return state;
  }
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.FETCH_PREDICTION_FAILURE:
    case actionTypes.FETCH_PREDICTION_DETAILS_FAILURE:
    case actionTypes.FETCH_ADDRESS_FAILURE:
    case actionTypes.FETCH_ADDRESSES_FAILURE:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_PREDICTION_REQUEST:
    case actionTypes.FETCH_PREDICTION_DETAILS_REQUEST:
    case actionTypes.FETCH_ADDRESS_REQUEST:
    case actionTypes.FETCH_ADDRESSES_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.FETCH_PREDICTION_REQUEST:
    case actionTypes.FETCH_PREDICTION_DETAILS_REQUEST:
    case actionTypes.FETCH_ADDRESSES_REQUEST:
    case actionTypes.FETCH_ADDRESS_REQUEST:
    case actionTypes.CREATE_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.FETCH_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return true;
    case actionTypes.FETCH_PREDICTION_FAILURE:
    case actionTypes.FETCH_PREDICTION_SUCCESS:
    case actionTypes.FETCH_PREDICTION_DETAILS_FAILURE:
    case actionTypes.FETCH_PREDICTION_DETAILS_SUCCESS:
    case actionTypes.FETCH_ADDRESSES_FAILURE:
    case actionTypes.FETCH_ADDRESSES_SUCCESS:
    case actionTypes.FETCH_ADDRESS_FAILURE:
    case actionTypes.FETCH_ADDRESS_SUCCESS:
    case actionTypes.CREATE_ADDRESS_FAILURE:
    case actionTypes.CREATE_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.FETCH_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.FETCH_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const predictions = (state = INITIAL_STATE.predictions, action = {}) => {
  switch (action.type) {
    case actionTypes.FETCH_PREDICTION_REQUEST:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: true,
      };
    case actionTypes.FETCH_PREDICTION_FAILURE:
      return {
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_PREDICTION_SUCCESS:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: false,
        result: action.payload,
      };
    case actionTypes.RESET_PREDICTION:
      return INITIAL_STATE.predictions;
    default:
      return state;
  }
};

export const predictionDetails = (
  state = INITIAL_STATE.predictionDetails,
  action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_PREDICTION_DETAILS_REQUEST:
      return {
        error: INITIAL_STATE.predictionDetails.error,
        isLoading: true,
      };
    case actionTypes.FETCH_PREDICTION_DETAILS_FAILURE:
      return {
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_PREDICTION_DETAILS_SUCCESS:
      return {
        error: INITIAL_STATE.predictionDetails.error,
        isLoading: false,
        result: action.payload,
      };
    case actionTypes.RESET_PREDICTION:
      return INITIAL_STATE.predictionDetails;
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.CREATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const createdAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...createdAddress },
      };
    });
  },
  [actionTypes.UPDATE_ADDRESS_SUCCESS]: (state, action) => {
    const id = action.payload.result;
    const updatedAddress = action.payload.entities.addresses[id];

    return produce(state, draftState => {
      if (!draftState.addresses) {
        draftState.addresses = {};
      }

      draftState.addresses = {
        ...draftState.addresses,
        [id]: { ...updatedAddress },
      };
    });
  },
  [actionTypes.REMOVE_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;
    const currentAddresses = state.addresses;

    return produce(state, draftState => {
      draftState.addresses = omit(currentAddresses, addressId);
    });
  },
  [actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevCurrentShippingAddress = getDefaultAddress(
      state.addresses,
      'isCurrentShipping',
    );

    return produce(state, draftState => {
      if (prevCurrentShippingAddress) {
        // Unmark previous shipping address as default
        draftState.addresses[
          prevCurrentShippingAddress.id
        ].isCurrentShipping = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isCurrentShipping = true;
    });
  },
  [actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevCurrentBillingAddress = getDefaultAddress(
      state.addresses,
      'isCurrentBilling',
    );

    return produce(state, draftState => {
      if (prevCurrentBillingAddress) {
        // Unmark previous billing address as default
        draftState.addresses[
          prevCurrentBillingAddress.id
        ].isCurrentBilling = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isCurrentBilling = true;
    });
  },
  [actionTypes.FETCH_ADDRESS_SCHEMA_SUCCESS]: (state, action) => {
    const countryId = action.payload.result;
    const countrySchema = get(
      action,
      `payload.entities.addressSchema[${countryId}]`,
      {},
    );

    return produce(state, draftState => {
      if (!draftState.addressSchema) {
        draftState.addressSchema = {};
      }

      draftState.addressSchema[countryId] = { ...countrySchema };
    });
  },
  [actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    // Get prev default address so it can later be unmarked as the default
    const prevCurrentContactAddress = getDefaultAddress(
      state.addresses,
      'isCurrentPreferred',
    );

    return produce(state, draftState => {
      if (prevCurrentContactAddress) {
        // Unmark previous contact address as default
        draftState.addresses[
          prevCurrentContactAddress.id
        ].isCurrentPreferred = false;
      }
      // Select the selected address as default
      draftState.addresses[addressId].isCurrentPreferred = true;
    });
  },
  [actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_SUCCESS]: (state, action) => {
    const { addressId } = action.meta;

    return produce(state, draftState => {
      // Unmark the selected address as default
      draftState.addresses[addressId].isCurrentPreferred = false;
    });
  },
  [LOGOUT_SUCCESS]: state => {
    const newState = { ...state };
    delete newState.addresses;

    return newState;
  },
};

export const address = (state = INITIAL_STATE.address, action = {}) => {
  switch (action.type) {
    case actionTypes.CREATE_ADDRESS_REQUEST:
      return {
        isLoading: { ...state.isLoading },
        error: {},
      };
    case actionTypes.FETCH_ADDRESS_REQUEST:
    case actionTypes.UPDATE_ADDRESS_REQUEST:
    case actionTypes.REMOVE_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_REQUEST:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_REQUEST:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: true,
        },
        error: {},
      };
    case actionTypes.FETCH_ADDRESS_SUCCESS:
    case actionTypes.UPDATE_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_SUCCESS:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_SUCCESS:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
      };
    case actionTypes.FETCH_ADDRESS_FAILURE:
    case actionTypes.UPDATE_ADDRESS_FAILURE:
    case actionTypes.REMOVE_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_BILLING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_SHIPPING_ADDRESS_FAILURE:
    case actionTypes.SET_DEFAULT_CONTACT_ADDRESS_FAILURE:
    case actionTypes.REMOVE_DEFAULT_CONTACT_ADDRESS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.addressId]: false,
        },
        error: {
          ...state.error,
          [action.meta.addressId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const addresses = reducerFactory(
  'FETCH_ADDRESSES',
  INITIAL_STATE.addresses,
  actionTypes,
);

export const addressSchema = reducerFactory(
  'FETCH_ADDRESS_SCHEMA',
  INITIAL_STATE.addressSchema,
  actionTypes,
);

export const defaultAddressDetails = createReducerWithResult(
  'FETCH_DEFAULT_CONTACT_ADDRESS',
  INITIAL_STATE.defaultAddressDetails,
  actionTypes,
);

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getResult = state => state.result;

export const getPredictions = state => state.predictions;
export const getPredictionDetails = state => state.predictionDetails;
export const getAddresses = state => state.addresses;
export const getAddress = state => state.address;
export const getAddressSchema = state => state.addressSchema;
export const getDefaultAddressDetails = state => state.defaultAddressDetails;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  predictions,
  predictionDetails,
  addresses,
  address,
  addressSchema,
  defaultAddressDetails,
});

/**
 * Reducer for addresses state.
 *
 * @function addressesReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (
    action.type === actionTypes.RESET_ADDRESSES ||
    action.type === LOGOUT_SUCCESS
  ) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};
