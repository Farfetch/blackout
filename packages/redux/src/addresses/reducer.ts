import * as actionTypes from './actionTypes.js';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import {
  FETCH_USER_SUCCESS,
  LOGIN_SUCCESS,
  LOGOUT_SUCCESS,
  REGISTER_SUCCESS,
} from '../users/authentication/actionTypes.js';
import type * as T from './types/index.js';

export const INITIAL_STATE: T.AddressesState = {
  predictions: {
    result: null,
    error: null,
    isLoading: false,
  },
  prediction: {
    result: null,
    error: null,
    isLoading: false,
  },
};

const predictions = (
  state = INITIAL_STATE.predictions,
  action: AnyAction,
): T.AddressesState['predictions'] => {
  switch (action.type) {
    case actionTypes.FETCH_ADDRESS_PREDICTIONS_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.predictions.error,
        isLoading: true,
      };
    case actionTypes.FETCH_ADDRESS_PREDICTIONS_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_ADDRESS_PREDICTIONS_SUCCESS:
      return {
        error: INITIAL_STATE.predictions.error,
        isLoading: false,
        result: action.payload,
      };
    default:
      return state;
  }
};

const prediction = (
  state = INITIAL_STATE.prediction,
  action: AnyAction,
): T.AddressesState['prediction'] => {
  switch (action.type) {
    case actionTypes.FETCH_ADDRESS_PREDICTION_DETAILS_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.prediction.error,
        isLoading: true,
      };
    case actionTypes.FETCH_ADDRESS_PREDICTION_DETAILS_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_ADDRESS_PREDICTION_DETAILS_SUCCESS:
      return {
        error: INITIAL_STATE.prediction.error,
        isLoading: false,
        result: action.payload,
      };
    default:
      return state;
  }
};

export const getAddressPredictions = (
  state: T.AddressesState,
): T.AddressesState['predictions'] => state.predictions;
export const getAddressPrediction = (
  state: T.AddressesState,
): T.AddressesState['prediction'] => state.prediction;

const reducer = combineReducers({
  predictions,
  prediction,
});

/**
 * Reducer for addresses state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const addressesReducer: Reducer<T.AddressesState> = (state, action) => {
  if (
    action.type === LOGOUT_SUCCESS ||
    action.type === LOGIN_SUCCESS ||
    action.type === REGISTER_SUCCESS ||
    action.type === FETCH_USER_SUCCESS ||
    action.type === actionTypes.RESET_ADDRESS_PREDICTIONS
  ) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};

export default addressesReducer;
