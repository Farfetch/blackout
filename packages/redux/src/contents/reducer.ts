import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import type {
  ActionFetchCommercePages,
  ActionFetchContent,
  ActionFetchContentTypes,
  ActionFetchSEO,
  State,
} from './types';
import type { ReducerSwitch } from '../types';

export const INITIAL_STATE_CONTENT: State = {
  searchResults: {},
  contentTypes: {
    error: undefined,
    isLoading: false,
    result: null,
  },
  metadata: {
    error: {},
    isLoading: {},
    result: null,
  },
};

const searchResults = (
  state = INITIAL_STATE_CONTENT.searchResults,
  action: ActionFetchContent | ActionFetchCommercePages,
): State['searchResults'] => {
  switch (action.type) {
    case actionTypes.FETCH_CONTENT_REQUEST:
    case actionTypes.FETCH_COMMERCE_PAGES_REQUEST:
      return {
        ...state,
        [action.payload.hash]: {
          error: null,
          isLoading: true,
        },
      };
    case actionTypes.FETCH_CONTENT_SUCCESS:
    case actionTypes.FETCH_COMMERCE_PAGES_SUCCESS:
      return {
        ...state,
        [action.payload.hash]: {
          isLoading: false,
          result: action.payload.result,
        },
      };
    case actionTypes.FETCH_CONTENT_FAILURE:
    case actionTypes.FETCH_COMMERCE_PAGES_FAILURE:
      return {
        ...state,
        [action.payload.hash]: {
          isLoading: undefined,
          error: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const contentTypes = (
  state = INITIAL_STATE_CONTENT.contentTypes,
  action: ActionFetchContentTypes,
): State['contentTypes'] => {
  switch (action.type) {
    case actionTypes.FETCH_CONTENT_TYPES_REQUEST:
      return {
        isLoading: true,
        error: INITIAL_STATE_CONTENT.contentTypes.error,
      };
    case actionTypes.FETCH_CONTENT_TYPES_SUCCESS:
      return {
        result: action.payload,
        isLoading: false,
      };
    case actionTypes.FETCH_CONTENT_TYPES_FAILURE:
      return {
        isLoading: false,
        error: action.payload.error,
      };
    default:
      return state;
  }
};

const metadata = (
  state = INITIAL_STATE_CONTENT.metadata,
  action: ActionFetchSEO,
): State['metadata'] => {
  switch (action.type) {
    case actionTypes.FETCH_SEO_REQUEST:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: true,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: null,
        },
      };
    case actionTypes.FETCH_SEO_SUCCESS:
      return {
        ...state,
        result: {
          ...state.result,
          [action.payload.pathname]: action.payload.result,
        },
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: false,
        },
      };
    case actionTypes.FETCH_SEO_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.pathname]: undefined,
        },
        error: {
          ...state.error,
          [action.payload.pathname]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const getContentResult = (state: State): State['searchResults'] =>
  state.searchResults;
export const getContentTypes = (state: State): State['contentTypes'] =>
  state.contentTypes;
export const getSEOmetadata = (state: State): State['metadata'] =>
  state.metadata;
export const getContentTypesError = (
  state: State,
): State['contentTypes']['error'] => state.contentTypes.error;

const reducers = combineReducers({
  searchResults,
  contentTypes,
  metadata,
});

/**
 * Reducer for contents state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const contentsReducer: ReducerSwitch<
  State,
  ActionFetchContent | ActionFetchContentTypes | ActionFetchSEO
> = (state, action): State => {
  if (action.type === actionTypes.RESET_CONTENTS) {
    return reducers(INITIAL_STATE_CONTENT, action);
  }

  return reducers(state, action);
};

export default contentsReducer;
