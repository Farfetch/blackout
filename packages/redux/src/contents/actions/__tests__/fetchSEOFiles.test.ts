import * as actionTypes from '../../actionTypes.js';
import { fetchSEOFiles } from '../index.js';
import { find } from 'lodash-es';
import { getSEOFiles } from '@farfetch/blackout-client';
import {
  hash,
  seoFilesData,
  seoFilesQuery,
} from 'tests/__fixtures__/contents/seoFiles.fixtures.mjs';
import { INITIAL_STATE_CONTENT } from '../../reducer.js';
import { mockStore } from '../../../../tests/index.js';

jest.mock('@farfetch/blackout-client', () => ({
  ...jest.requireActual('@farfetch/blackout-client'),
  getSEOFiles: jest.fn(),
}));

const contentsSEOFilesMockStore = (state = {}) =>
  mockStore({ contents: INITIAL_STATE_CONTENT }, state);

const expectedConfig = undefined;
let store: ReturnType<typeof contentsSEOFilesMockStore>;

describe('feachSEOFiles action create', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    store = contentsSEOFilesMockStore();
  });

  it('should create the correct actions for when featch SEO procedute fails', async () => {
    const expectedError = new Error('Get SEO Files error');

    (getSEOFiles as jest.Mock).mockRejectedValueOnce(expectedError);

    await expect(
      async () => await fetchSEOFiles(seoFilesQuery)(store.dispatch),
    ).rejects.toThrow(expectedError);

    expect(getSEOFiles).toHaveBeenCalledTimes(1);
    expect(getSEOFiles).toHaveBeenCalledWith(seoFilesQuery, expectedConfig);
    expect(store.getActions()).toEqual([
      {
        payload: { hash },
        type: actionTypes.FETCH_SEO_FILES_REQUEST,
      },
      {
        payload: { error: expectedError, hash },
        type: actionTypes.FETCH_SEO_FILES_FAILURE,
      },
    ]);
  });

  it('should create the correct actions for when the fetch SEO Files procedure is successful', async () => {
    (getSEOFiles as jest.Mock).mockResolvedValueOnce(seoFilesData);

    await fetchSEOFiles(seoFilesQuery)(store.dispatch).then(clientResult => {
      expect(clientResult).toBe(seoFilesData);
    });

    const actionResults = store.getActions();

    expect(getSEOFiles).toHaveBeenCalledWith(seoFilesQuery, expectedConfig);
    expect(actionResults).toEqual([
      {
        payload: { hash },
        type: actionTypes.FETCH_SEO_FILES_REQUEST,
      },
      {
        payload: { result: seoFilesData, hash },
        type: actionTypes.FETCH_SEO_FILES_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, { type: actionTypes.FETCH_SEO_FILES_SUCCESS }),
    ).toMatchSnapshot('Get SEO Files payload');
  });
});
