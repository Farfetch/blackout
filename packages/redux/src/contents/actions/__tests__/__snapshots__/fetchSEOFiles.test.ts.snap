// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feachSEOFiles action create should create the correct actions for when the fetch SEO Files procedure is successful: Get SEO Files payload 1`] = `
Object {
  "payload": Object {
    "hash": "siteSEOFiles!1234",
    "result": Object {
      "entries": Array [
        Object {
          "content": "string",
          "hostId": 1234,
          "name": "siteSEOFiles.txt",
          "path": "siteSEOFiles.txt",
          "subfolderStructure": "/en-US",
          "type": "Pages",
          "uploadDate": "2023-05-11T15:56:05.712Z",
        },
      ],
      "number": 0,
      "totalItems": 1,
      "totalPages": 1,
    },
  },
  "type": "@farfetch/blackout-redux/FETCH_SEO_FILES_SUCCESS",
}
`;
